src/Sylius/Bundle/PromotionBundle/Form/Type/CatalogPromotionActionType.php

```diff
@@ -48,18 +48,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
 
         $builder
             ->addEventListener(FormEvents::PRE_SET_DATA, function(FormEvent $event): void {
-                /** @var CatalogPromotionActionInterface|null $data */
-                $data = $event->getData();
-                if ($data === null) {
-                    return;
-                }
-
-                $form = $event->getForm();
-
-                $actionConfigurationType = $this->actionConfigurationTypes[$data->getType()];
-                $form->add('configuration', $actionConfigurationType, [
-                    'label' => false,
-                ]);
+               $this->addTypeToForm($event);
             })
             ->addEventListener(FormEvents::PRE_SUBMIT, function(FormEvent $event): void {
```

Comment:
i guess the problem was with this EventListener as its kinda big, maybe we could move some logic to private method to understand what's going on?

Reply:
@AdamKasp have you checked if all 3 listeners are needed?

Reply:
@SirDomin TBH I Don't see good way to exclude sth to private method,
@GSadee I've checked it :)